src/main/java/org/folio/rest/impl/InvoiceLineHelper.java

```diff
@@ -156,20 +172,29 @@ private void updateOutOfSyncInvoiceLine(InvoiceLine invoiceLine) {
 
   public CompletableFuture<Void> updateInvoiceLine(InvoiceLine invoiceLine) {
 
+    // 1. Get invoice line from storage
     return getInvoiceLine(invoiceLine.getId())
-      .thenCompose(existedInvoiceLine -> new InvoiceHelper(okapiHeaders, ctx, lang).getInvoiceRecord(existedInvoiceLine.getInvoiceId())
-        .thenAccept(existedInvoice -> {
-          validateInvoiceLine(existedInvoice, invoiceLine, existedInvoiceLine);
-          invoiceLine.setInvoiceLineNumber(existedInvoiceLine.getInvoiceLineNumber());
-          HelperUtils.calculateInvoiceLineTotals(invoiceLine, existedInvoice);
-        })
-      )
-      .thenCompose(ok -> updateInvoiceLineToStorage(invoiceLine))
-      .thenAccept(ok -> updateInvoice(invoiceLine.getInvoiceId()));
+      // 2. Get invoice from storage
+      .thenCompose(invoiceLineFromStorage -> getInvoice(invoiceLineFromStorage).thenCompose(invoice -> {
+        // 3. Validate if invoice line update is allowed
+        validateInvoiceLine(invoice, invoiceLine, invoiceLineFromStorage);
+        invoiceLine.setInvoiceLineNumber(invoiceLineFromStorage.getInvoiceLineNumber());
+
+        // 4. Recalculate totals before update which also indicates if invoice requires update
+        calculateInvoiceLineTotals(invoiceLine, invoice);
+
+        // 5. Update invoice line in storage
+        return updateInvoiceLineToStorage(invoiceLine).thenRun(() -> {
```

Comment:
Will invoiceLine be updated to storage before `calculateInvoiceLineTotals` even completes?

Reply:
`calculateInvoiceLineTotals` returns `InvoiceLine` which may be updated with actual totals. This is blocking operation i.e. `updateInvoiceLineToStorage` can start only when `calculateInvoiceLineTotals` completed its work

Reply:
I do not see any error messages in the console with respect to this operation so I think it should be fine. We have handled it asynchronously in other places so that’s why I was curious https://github.com/folio-org/mod-invoice/blob/master/src/main/java/org/folio/rest/impl/InvoiceLineHelper.java#L137